source/text.c

```diff
@@ -249,24 +350,217 @@ void C2D_DrawText(const C2D_Text* text, u32 flags, float x, float y, float z, fl
+					glyphY = y + dispY*words[consecutiveWordNum].newLineNumber;
+				}
+				else
+				{
+					glyphX = x + scaleX*(cur->xPos - finalLineWidths[cur->lineNo]);
+					glyphY = y + dispY*cur->lineNo;
+				}
+
+				C2Di_SetTex(cur->sheet);
+				C2Di_Update();
+				C2Di_AppendVtx(glyphX,        glyphY,        glyphZ, cur->texcoord.left,  cur->texcoord.top,    0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX,        glyphY+glyphH, glyphZ, cur->texcoord.left,  cur->texcoord.bottom, 0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX+glyphW, glyphY,        glyphZ, cur->texcoord.right, cur->texcoord.top,    0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX+glyphW, glyphY,        glyphZ, cur->texcoord.right, cur->texcoord.top,    0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX,        glyphY+glyphH, glyphZ, cur->texcoord.left,  cur->texcoord.bottom, 0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX+glyphW, glyphY+glyphH, glyphZ, cur->texcoord.right, cur->texcoord.bottom, 0.0f, 1.0f, color);
+			}
+		}
+		break;
+		case C2D_AlignCenter:
+		{
+			float finalLineWidths[flags & C2D_WordWrap ? words[text->words-1].newLineNumber + 1 : text->lines];
+			C2Di_CalcLineWidths(finalLineWidths, text, words, flags & C2D_WordWrap);
+
+			for (cur = begin; cur != end; cur++)
+			{
+				float glyphW = scaleX*cur->width;
+				float glyphX;
+				float glyphY;
+				if (flags & C2D_WordWrap)
+				{
+					u32 consecutiveWordNum = cur->wordNo + lines[cur->lineNo].wordStart;
+					glyphX = x + scaleX*(cur->xPos + words[consecutiveWordNum].wrapXOffset - finalLineWidths[words[consecutiveWordNum].newLineNumber]/2);
+					glyphY = y + dispY*words[consecutiveWordNum].newLineNumber;
+				}
+				else
+				{
+					glyphX = x + scaleX*(cur->xPos - finalLineWidths[cur->lineNo]/2);
+					glyphY = y + dispY*cur->lineNo;
+				}
+
+				C2Di_SetTex(cur->sheet);
+				C2Di_Update();
+				C2Di_AppendVtx(glyphX,        glyphY,        glyphZ, cur->texcoord.left,  cur->texcoord.top,    0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX,        glyphY+glyphH, glyphZ, cur->texcoord.left,  cur->texcoord.bottom, 0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX+glyphW, glyphY,        glyphZ, cur->texcoord.right, cur->texcoord.top,    0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX+glyphW, glyphY,        glyphZ, cur->texcoord.right, cur->texcoord.top,    0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX,        glyphY+glyphH, glyphZ, cur->texcoord.left,  cur->texcoord.bottom, 0.0f, 1.0f, color);
+				C2Di_AppendVtx(glyphX+glyphW, glyphY+glyphH, glyphZ, cur->texcoord.right, cur->texcoord.bottom, 0.0f, 1.0f, color);
+			}
+		}
+		break;
+		case C2D_AlignJustified:
+		{
+			if (!(flags & C2D_WordWrap))
+			{
+				lines = alloca(sizeof(*lines)*text->lines);
+				words = alloca(sizeof(*words)*text->words);
+				C2Di_CalcLineInfo(text, lines, words);
+			}
+			// Get total width available for whitespace for all lines after wrapping
+			struct justifyInfo
```

Comment:
You can do this instead of defining a temporary type:

```c
struct
{
    float whitespaceWidth;
    u32 wordStart;
    u32 words;
} justifiedLineInfo[words[text->words - 1].newLineNumber + 1];
```